Introduce BlobId V3 and the idea of native and crafted ids

pnarayanan opened the pull request:
Introduce V3 of the BlobId.

Introduce a type for blobs. For V3:
  - 'Native' blobs are those that are created by the router in the
    context of a PUT operation.
  - 'Crafted' blobs are those that are V3 blobs crafted outside of
    the router, such as say, converted V2 or V1 ids.

Take the current active blob id version from a config setting in
RouterConfig to make enabling new versions easy.

Tests pass with the config value set to 2 as well as 3.

vgkholla commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -353,21 +394,36 @@ public boolean equals(Object o) {
   @Override
   public int hashCode() {
-    return Utils.hashcode(new Object[]{version, partitionId, uuid});
+    return Utils.hashcode(new Object[]{uuid});
```
Are you sure this change is safe? Shouldn't this be based on versions too?

cgtz replied:
Could we just do `uuid.hashcode()` directly here?

vgkholla commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -353,21 +394,36 @@ public boolean equals(Object o) {
+  /**
+   * @return all valid versions of BlobId.
+   */
+  public static Short[] getAllValidVersions() {
```
decision b/w `short` and `Short`? I prefer `short` here since this cannot be `null`

vgkholla replied:
Also does this need to be `public`?

vgkholla replied:
Seems like it's only used in tests now - do you envision a production use case for this?

vgkholla commented on ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java:
```diff
@@ -425,19 +340,12 @@ private void assertBlobIdLongForm(short version) throws IOException {
-    }
+    return new BlobId(version, type, datacenterId, accountId, containerId, partitionId);
   }
 }
```
any tests for the `compareTo()` method to ensure that only uuid is the one being compared ?

vgkholla commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -132,6 +133,8 @@ public AmbryBlobStorageServiceTest() throws Exception {
     RestRequestMetricsTracker.setDefaults(metricRegistry);
     configProps.setProperty("frontend.allow.service.id.based.post.request",
         String.valueOf(shouldAllowServiceIdBasedPut));
+    configProps.setProperty("router.hostname", "localhost");
```
is this required?

vgkholla commented on ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java:
```diff
@@ -474,55 +478,58 @@ public void injectionAccountAndContainerForPutTest() throws Exception {
+      verifyAccountAndContainerFromBlobId(blobId, null, null, RestServiceErrorCode.InvalidAccount);
+
+      // aid=nonExistAId, cid=nonExistCId
+      blobId = new BlobId(version, BlobId.BlobIdType.NATIVE, ClusterMapUtils.UNKNOWN_DATACENTER_ID, (short) -1234,
```
i'm assuming no actual code in the loop was changed. Please let me know if it is otherwise.

cgtz commented on ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java:
```diff
@@ -81,34 +80,26 @@ public BlobIdTest(short version) throws Exception {
+        referencePartitionId);
+    deserializeBlobIdAndAssert(version, blobId.getID());
   }
 
```
Could you add a test for hashcode, and compareto within the same version?

cgtz replied:
true, not much to test for hashcode. What about compareTo within the same version? I think that's useful.

pnarayanan replied:
Yes and I actually added intra-version comparisons in the `testComparisons()` method (and renamed it from `testInterVersionComparison()`) as you can see below.

I am also adding a few more asserts in the SerDe test in the next commit to ensure that blob after serialization and deserialization is exactly the same as the original as far as comparisons are concerned (`equals`, `compareTo` and `hashCode()` all match exactly).

cgtz commented on ambry-protocol/src/test/java/com.github.ambry.protocol/MessageInfoAndMetadataListSerDeTest.java:
```diff
@@ -58,10 +59,12 @@ public void testSerDe() throws Exception {
-            partitionId), new BlobId((byte) 0, (byte) 0, accountIds[3], containerIds[3], partitionId)};
+    short blobIdVersion = CommonTestUtils.getCurrentBlobIdVersion();
+    StoreKey[] keys = {new BlobId(blobIdVersion, BlobId.BlobIdType.NATIVE, (byte) 0, accountIds[0], containerIds[0],
+        partitionId), new BlobId(blobIdVersion, BlobId.BlobIdType.NATIVE, (byte) 0, accountIds[1], containerIds[1],
```
should we test with multiple blob ID versions in the same list?

cgtz commented on ambry-replication/src/test/java/com.github.ambry.replication/ReplicationTest.java:
```diff
@@ -320,9 +321,11 @@ public void replicaThreadTest() throws Exception {
 
       short accountId = Utils.getRandomShort(TestUtils.RANDOM);
       short containerId = Utils.getRandomShort(TestUtils.RANDOM);
+      short blobIdVersion = CommonTestUtils.getCurrentBlobIdVersion();
```
Could this be a class member or static variable?

cgtz commented on ambry-router/src/test/java/com.github.ambry.router/InMemoryRouter.java:
```diff
@@ -331,9 +332,9 @@ public void run() {
+          ClusterMapUtils.UNKNOWN_DATACENTER_ID, postData.getBlobProperties().getAccountId(),
+          postData.getBlobProperties().getContainerId(), getPartitionForPut()).getID();
       if (blobs.containsKey(blobId)) {
         exception = new RouterException("Blob ID duplicate created.", RouterErrorCode.UnexpectedInternalError);
```
I added some code in #792 to allow puts to InMemoryRouter with a specific blob ID version. I suppose that can be replaced by an `InMemoryRouter.setBlobIdVersion()` method.

pnarayanan replied:
Haven't looked at that, how are you setting a specific version now? Will talk to you.

cgtz replied:
I was using BlobIdV1, which overrides the getVersion method

cgtz commented on ambry-tools/src/main/java/com.github.ambry/tools/perf/ServerWritePerformance.java:
```diff
@@ -320,6 +322,10 @@ public ServerWritePerfRun(int threadIndex, Throttler throttler, AtomicBoolean is
+      Properties props = new Properties();
+      props.setProperty("router.hostname", "localhost");
+      props.setProperty("router.datacenter.name", "localDC");
+      blobIdVersion = new RouterConfig(new VerifiableProperties(props)).routerBlobIdCurrentVersion;
```
I wonder if the default could just be stored as a constant inside of `RouterConfig`, instead of having to construct properties objects in all of these places.

pnarayanan replied:
This change is not very extensive (only one place in the production code creates a `BlobId`). Rest of the changes are in tests (for which we created a method in `CommonTestUtils`) and these are the tools.

There is a problem with keeping it as a constant in `RouterConfig` (basically consumers of it are interested in actuals not defaults). We could potentially use a `static` method, but then there are initialization order problems to worry about.

cgtz commented on ambry-api/src/main/java/com.github.ambry/config/RouterConfig.java:
```diff
@@ -147,6 +147,13 @@
+   * The version to use for new BlobIds.
+   */
+  @Config("router.blobid.current.version")
+  @Default("2")
```
Consider just making the default value a static final field in this config class. That would get rid of the need for CommonTestUtils.

pnarayanan commented:
@cgtz @vgkholla Addressed comments. Will discuss some of the points in person. (Build failure is unrelated unfortunately, have initiated another one).

cgtz commented on ambry-api/src/main/java/com.github.ambry/config/VerifiableProperties.java:
```diff
@@ -122,6 +124,24 @@ public Short getShortInRange(String name, Short defaultVal, Short start, Short e
+   * @param allowedValues The array of allowed values for this property.
+   * @return the Short value
+   */
+  public short getShortFromAllowedValues(String name, Short defaultVal, Short[] allowedValues) {
```
I believe you can pass in an array when you have ... arg, but I might be mistaken

pnarayanan commented on ambry-commons/src/test/java/com.github.ambry.commons/CommonTestUtils.java:
```diff
@@ -0,0 +1,35 @@
+public class CommonTestUtils {
+  /**
+   * Get the current active blob id version via {@link RouterConfig}
+   * @return the current active blbo id version.
```
fixed.

pnarayanan commented:
@vgkholla @cgtz Addressed your comments.

vgkholla commented on ambry-api/src/main/java/com.github.ambry/config/RouterConfig.java:
```diff
@@ -147,6 +147,13 @@
+   */
+  @Config("router.blobid.version")
+  @Default("2")
+  public final short routerBlobidCurrentVersion;
```
It's not a config test I'm suggesting - there doesn't seem to be a test in `PutOperation` where you change the config and it writes in the version you want it to i.e. something that would fail if the the way of generating `BlobId` in `PutOperation` if the code had never been changed.

pnarayanan replied:
For an incremental change such as this it might appear that we want such a test, but we don't generally have tests for whether the config values are respected by the components. Tests do change certain configs - but generally they do that to test the functionality of the flow enabled by that change, not to test whether such a flow gets enabled at all.

For example, if we wanted to test whether the router would have any difficulty putting blobs in V3 (which I don't think we need to - Blob id construction tests suffice), then we would set the config accordingly - but not to test whether by setting the config the blob id that the router creates is actually a V3 blob id.

vgkholla commented:
@pnarayanan changes LGTM. Have left a comment about a test I would like us to have. Please see if that is possible (i'm not sure)

cgtz commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -353,21 +397,35 @@ public boolean equals(Object o) {
    */
-  protected short getCurrentVersion() {
-    return CURRENT_VERSION;
+  protected short getVersion() {
```
Does this method still need to protected since we are passing in the version in the constructor?

pnarayanan replied:
This method is not being used anymore, removed it for now.

cgtz commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -353,21 +394,36 @@ public boolean equals(Object o) {
+   * @return all valid versions of BlobId.
+   */
+  public static Short[] getAllValidVersions() {
+    return new Short[]{BLOB_ID_V1, BLOB_ID_V2, BLOB_ID_V3};
```
That is true. Also, I guess there's the concern that arrays are mutable, so what you have is probably better.

cgtz commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -353,21 +397,35 @@ public boolean equals(Object o) {
+    /**
+     * Indicates natively created BlobId (in the context of a PUT operation)
+     */
+    NATIVE, /**
```
The autoformatter seems to have joined these lines. I usually just leave a blank line in between to satisfy it.

pnarayanan replied:
Ah, didn't know adding a new line fixes this. Done.

pnarayanan commented:
@cgtz Addressed your comments.
